File: profiling/utils/firecracker_vm.py
import subprocess
import time

import utils.tmux as tmux
import utils.constants as constants
import utils.uffd as uffd



class FirecrackerVM():
    def __init__(self, session_name: str) -> None:
        self.session_name = session_name
        self.tmux_client = tmux.Tmux()
        self.uffd_client = uffd.Uffd()

    def start_fc_vmm(self) -> None:
        subprocess.run(f"bash {constants.SCRIPTS_PATH}/fc_vmm.sh", shell=True)

    def stop_fc_vmm(self) -> None:
        subprocess.run(f"tmux kill-session -t fc_vmm", shell=True)

    def create_vm(self) -> None:
        subprocess.run(f"bash {constants.SCRIPTS_PATH}/run_fc.sh", shell=True)

    def stop_vm(self) -> None:
        subprocess.run(f"ssh -i {constants.UBUBTU_PATH}/ubuntu-22.04.id_rsa root@172.16.0.2 'reboot'", shell=True)
        time.sleep(15)

    def run_command_on_vm(self, command: str) -> None:
        subprocess.run(f"ssh -i {constants.UBUBTU_PATH}/ubuntu-22.04.id_rsa root@172.16.0.2 '{command}'", shell=True)
    
    def run_command_on_vm_tmux(self, command: str) -> None:
        subprocess.run(f"ssh -i {constants.UBUBTU_PATH}/ubuntu-22.04.id_rsa root@172.16.0.2 tmux kill-session -t server", shell=True)
        time.sleep(1)
        subprocess.run(f"ssh -i {constants.UBUBTU_PATH}/ubuntu-22.04.id_rsa root@172.16.0.2 tmux new -s server -d", shell=True)
        time.sleep(1)
        subprocess.run(f"ssh -i {constants.UBUBTU_PATH}/ubuntu-22.04.id_rsa root@172.16.0.2 tmux send -t server '\"{command}\"' ENTER", shell=True)
    
    def snapshot_vm(self) -> None:
        subprocess.run(f"bash {constants.SCRIPTS_PATH}/snapshot_capture.sh", shell=True)
    
    def restore_vm(self) -> None:
        subprocess.run(f"bash {constants.SCRIPTS_PATH}/snapshot_restore.sh", shell=True)
    
    def page_count(self) -> int:
        return self.uffd_client.get_uffd_pages()

File: utils/firecracker_vm.go
package utils

import (
	"fmt"
    "log"
    "os/exec"
    "time"
)

// FirecrackerVM struct to manage the Firecracker VM
type FirecrackerVM struct {
    SessionName  string
    TmuxClient   Tmux
    UffdClient   Uffd
}

// StartFCVMM starts the Firecracker VMM by running the fc_vmm.sh script
func (f *FirecrackerVM) StartFCVMM() {
    log.Println("Starting Firecracker VMM")
    cmd := exec.Command("bash", ScriptsPath+"/fc_vmm.sh")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to start VMM: %v", err)
    }
}

// StopFCVMM stops the Firecracker VMM by killing the tmux session
func (f *FirecrackerVM) StopFCVMM() {
    log.Println("Stopping Firecracker VMM")
    cmd := exec.Command("tmux", "kill-session", "-t", "fc_vmm")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to stop VMM: %v", err)
    }
}

// CreateVM runs the script to create the Firecracker VM
func (f *FirecrackerVM) CreateVM() {
    log.Println("Creating Firecracker VM")
    cmd := exec.Command("bash", ScriptsPath+"/run_fc.sh")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to create VM: %v", err)
    }
}

// StopVM stops the Firecracker VM by sending a reboot command via SSH
func (f *FirecrackerVM) StopVM() {
    log.Println("Stopping Firecracker VM")
    cmd := exec.Command("ssh", "-i", UbuntuPath+"/ubuntu-22.04.id_rsa", "root@192.168.0.2", "reboot")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to stop VM: %v", err)
    }
    time.Sleep(15 * time.Second)
}

// RunCommandOnVM runs a command inside the VM via SSH
func (f *FirecrackerVM) RunCommandOnVM(command string) {
    log.Printf("Running command on VM: %s", command)
    cmd := exec.Command("ssh", "-i", UbuntuPath+"/ubuntu-22.04.id_rsa", "root@192.168.0.2", command)
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to run command on VM: %v", err)
    }
}

// RunCommandOnVMTmux runs a command in tmux on the VM via SSH
func (f *FirecrackerVM) RunCommandOnVMTmux(command string) {
    log.Printf("Running tmux command on VM: %s", command)
    exec.Command("ssh", "-i", UbuntuPath+"/ubuntu-22.04.id_rsa", "root@192.168.0.2", "tmux kill-session -t server").Run()
    time.Sleep(1 * time.Second)
    exec.Command("ssh", "-i", UbuntuPath+"/ubuntu-22.04.id_rsa", "root@192.168.0.2", "tmux new -s server -d").Run()
    time.Sleep(1 * time.Second)
    exec.Command("ssh", "-i", UbuntuPath+"/ubuntu-22.04.id_rsa", "root@192.168.0.2", "tmux send -t server", fmt.Sprintf(`"%s"`, command), "ENTER").Run()
}

// SnapshotVM captures a snapshot of the VM using a script
func (f *FirecrackerVM) SnapshotVM() {
    log.Println("Capturing VM snapshot")
    cmd := exec.Command("bash", ScriptsPath+"/snapshot_capture.sh")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to capture snapshot: %v", err)
    }
}

// RestoreVM restores the VM from a snapshot using a script
func (f *FirecrackerVM) RestoreVM() {
    log.Println("Restoring VM snapshot")
    cmd := exec.Command("bash", ScriptsPath+"/snapshot_restore.sh")
    if err := cmd.Run(); err != nil {
        log.Fatalf("Failed to restore snapshot: %v", err)
    }
}

// PageCount retrieves the current UFFD page count using the Uffd client
func (f *FirecrackerVM) PageCount() int {
	num, _ := f.UffdClient.GetUffdPages()
    return num
}
